testsuite/fuzzer/fuzz/fuzz_targets/move/utils/vm.rs

```diff
@@ -149,20 +303,98 @@ pub(crate) fn check_for_invariant_violation(e: VMStatus) {
     }
 }
 
+// For VMError (verifier) invariants.
+const KNOWN_FALSE_POSITIVES_VMERROR: &[&str] =
+    &["too many type parameters/arguments in the program"];
+
+pub(crate) fn check_for_invariant_violation_vmerror(e: VMError) {
+    if e.status_type() == StatusType::InvariantViolation {
+        let is_known_false_positive = e.message().is_some_and(|msg| {
+            KNOWN_FALSE_POSITIVES_VMERROR
+                .iter()
+                .any(|known| msg.starts_with(known))
+        });
+
+        if !is_known_false_positive {
+            panic!(
+                "invariant violation {:?}\n{}{:?} {}",
+                e,
+                "RUST_BACKTRACE=1 DEBUG_VM_STATUS=",
+                e.major_status(),
+                "./fuzz.sh run move_aptosvm_publish_and_run <ARTIFACT>"
+            );
+        }
+    }
+}
+
+// Optional block execution wrapper — useful for consistent error mapping
+#[inline]
+pub(crate) fn execute_block_or_keep(
+    vm: &FakeExecutor,
+    block: Vec<aptos_types::transaction::SignedTransaction>,
+) -> Result<Vec<TransactionOutput>, Corpus> {
+    vm.execute_block(block).map_err(|e| {
+        check_for_invariant_violation(e);
+        Corpus::Keep
+    })
+}
+
+#[inline]
+pub(crate) fn sign_single_or_multi(
+    raw_tx: RawTransaction,
+    sender_acc: &Account,
+    secondary_addrs: &[move_core_types::account_address::AccountAddress],
+    accounts_by_addr: &HashMap<move_core_types::account_address::AccountAddress, Account>,
+) -> Result<SignedTransaction, Corpus> {
+    if secondary_addrs.is_empty() {
+        return raw_tx
+            .sign(
+                &sender_acc.privkey,
+                sender_acc.pubkey.as_ed25519().expect("ed25519 public key"),
+            )
+            .map_err(|_| Corpus::Reject)
+            .map(|c| c.into_inner());
+    }
+
+    let secondary_accounts: Vec<_> = secondary_addrs
+        .iter()
+        .map(|a| accounts_by_addr.get(a).expect("secondary account exists"))
+        .collect();
+    let secondary_privs = secondary_accounts.iter().map(|a| &a.privkey).collect();
+    raw_tx
+        .sign_multi_agent(
+            &sender_acc.privkey,
+            secondary_addrs.to_vec(),
+            secondary_privs,
+        )
+        .map_err(|_| Corpus::Reject)
+        .map(|c| c.into_inner())
+}
+
+// Shared sequential block metadata across all cached publishes and executions in this fuzzer.
+// Ensures AptosModuleCacheManager sees consecutive (parent, child) pairs and preserves caches.
+static NEXT_BLOCK_ID: AtomicU64 = AtomicU64::new(1);
+
+#[inline]
+pub(crate) fn next_block_metadata() -> TransactionSliceMetadata {
+    let child = NEXT_BLOCK_ID.fetch_add(1, Ordering::Relaxed);
```

Comment:
Where is this used? If this runs concurrently you need to use a stronger memory ordering for consistency